edgehog-device-runtime-forwarder/src/connections.rs

```diff
@@ -0,0 +1,362 @@
+                debug!("received ping, sending pong");
+                let msg = TungMessage::Pong(data);
+                self.send_to_ws(msg).await?;
+            }
+            TungMessage::Pong(_) => debug!("received Pong frame"),
+            TungMessage::Close(close_frame) => {
+                debug!("websocket close frame {close_frame:?}");
+                self.disconnect();
+                info!("closed every connection");
+                return Ok(ControlFlow::Break(()));
+            }
+            // text frames should never be sent
+            TungMessage::Text(data) => warn!("received Text websocket frame, {data}"),
+            TungMessage::Binary(bytes) => {
+                let res = match ProtoMessage::decode(&bytes) {
+                    // handle the actual protocol message
+                    Ok(proto_msg) => self.handle_proto_msg(proto_msg),
+                    Err(err) => {
+                        error!("failed to decode protobuf message due to {err:?}");
+                        Ok(())
+                    }
+                };
+
+                if let Err(err) = res {
+                    error!("failed to handle protobuf message due to {err:?}");
+                }
+            }
+            // wrong Message type
+            TungMessage::Frame(_) => error!("unhandled message type: {msg:?}"),
+        }
+
+        Ok(ControlFlow::Continue(()))
+    }
+
+    /// Handle a [`protobuf message`](ProtoMessage).
+    pub(crate) fn handle_proto_msg(&mut self, proto_msg: ProtoMessage) -> Result<(), Error> {
+        // remove from the collection all the terminated connections
+        self.connections.remove_terminated_connections();
+
+        // handle only HTTP requests, not other kind of protobuf messages
+        let http_req = match proto_msg.protocol {
+            ProtoProtocol::Http(Http::Request(http_req)) => http_req,
+            ProtoProtocol::Http(Http::Response(_http_res)) => {
+                error!("Http response should not be sent by the bridge");
+                return Err(Error::WrongMessage);
+            }
+            ProtoProtocol::WebSocket(_ws) => {
+                error!("WebSocket messages are not supported yet");
+                return Err(Error::WrongMessage);
+            }
+        };
+
+        self.connections.add_http(http_req)?;
+
+        Ok(())
+    }
+
+    /// Try to establish again a WebSocket connection with the bridge in case the connection is lost.
+    #[instrument(skip_all)]
+    pub(crate) async fn reconnect(&mut self) -> Result<(), Error> {
+        debug!("trying to reconnect");
+
+        self.ws_stream = Self::ws_connect(&self.url).await?;
+
+        info!("reconnected");
+
+        Ok(())
+    }
+
+    /// Close all the connections the device has established (e.g., with TTYD).
+    #[instrument(skip_all)]
+    pub(crate) fn disconnect(&mut self) {
+        info!("closing all the connections");
+        self.connections.disconnect();
+    }
+}
```

Comment:
I kept this struct and the respective methods just because everything in this file was related to the connections management. Also because the other structures in `connections.rs` are the errors and `Connections` (which are slightly small). Anyway, it seems also reasonable to move it into a separate file. What do you think?

Reply:
Yes, I would prefer to move it to a separate file.